gobblin-kafka/src/main/java/gobblin/kafka/writer/KafkaDataWriter.java

```diff
@@ -104,23 +153,68 @@ private static Properties stripPrefix(Properties props, String prefix) {
   @Override
   public void write(D record)
       throws IOException {
-    this.producer.send(new ProducerRecord<String, D>(topic, record), this.producerCallback);
+    log.debug("Write called");
+    lock.lock();
+    try
+    {
+      this.producer.send(new ProducerRecord<String, D>(topic, record), this.producerCallback);
+      recordsProduced++;
+    }
+    finally {
+      lock.unlock();
+    }
   }
 
   @Override
   public void close()
       throws IOException {
+    log.debug("Close called");
     this.producer.close();
   }
 
   @Override
   public void commit()
       throws IOException {
+    log.debug("Commit called, will wait for commitTimeout : " + commitTimeoutInNanos / MILLIS_TO_NANOS + "ms");
+    long commitStartTime = System.nanoTime();
+    lock.lock();
+    try {
+      while (((System.nanoTime() - commitStartTime) < commitTimeoutInNanos)  &&
+      (recordsProduced != (recordsWritten.get() + recordsFailed)))
+      {
+        log.debug("Commit waiting... records produced: " + recordsProduced + " written: " + recordsWritten.get() + " failed: " + recordsFailed);
+        try {
+          stateChange.awaitNanos(commitTimeoutInNanos);
+        } catch (InterruptedException e) {
+          throw new IOException(e);
+        }
+      }
+      log.debug("Commit done waiting");
+      if (recordsProduced != (recordsWritten.get() + recordsFailed)) // timeout
+      {
+        String message = "Commit timeout: RecordsWritten = " + recordsWritten.get()
+            + "; RecordsFailed = " + recordsFailed;
+        log.debug(message);
+        throw new IOException(message);
```

Comment:
If a writer fails to publish all messages, it will cause a failure of the work unit, so the watermark won't get advanced. This would cause Gobblin to keep resending the same messages over and over if even a single one is bad. Could we have different semantics here, where failures of a few messages still cause success? (We have metrics to notify user that there were failed messages).